python/pyarrow/tests/test_udf.py

```diff
@@ -504,3 +504,112 @@ def test_input_lifetime(unary_func_fixture):
     # Calling a UDF should not have kept `v` alive longer than required
     v = None
     assert proxy_pool.bytes_allocated() == 0
+
+
+def _record_batch_from_iters(schema, *iters):
+    arrays = [pa.array(list(v), type=schema[i].type)
+              for i, v in enumerate(iters)]
+    return pa.RecordBatch.from_arrays(arrays=arrays, schema=schema)
+
+
+def _record_batch_for_range(schema, n):
+    return _record_batch_from_iters(schema,
+                                    range(n, n + 10),
+                                    range(n + 1, n + 11))
+
+
+def make_udt_func(schema, batch_gen):
+    def udf_func(ctx):
+        class UDT:
+            def __init__(self):
+                self.caller = None
+
+            def __call__(self, ctx):
+                try:
+                    if self.caller is None:
+                        self.caller, ctx = batch_gen(ctx).send, None
+                    batch = self.caller(ctx)
+                except StopIteration:
+                    arrays = [pa.array([], type=field.type)
+                              for field in schema]
+                    batch = pa.RecordBatch.from_arrays(
+                        arrays=arrays, schema=schema)
```

Comment:
The reason is that, with the current design here, a UDTF must return an array (see [this post](https://github.com/apache/arrow/pull/14682/#discussion_r1045132971)). I don't know whether it's a good enough reason... If we'd like to allow a `StopIteration`, then I'd prefer this be done in a separate issue.

Reply:
I agree with much of what @westonpace writes here.

> What you've done here is use regular UDFs and allow the "init" function to be "given this input create a generator" and the "execute function" to be "return the next batch of data from the generator".

There are two main views for what's going on in this PR. One view is that a tabular UDF is not a scalar UDF, and they should be modeled separately, and hence this PR is abusing the scalar UDF construction. A second view is that both a scalar UDF and a tabular UDF are forms of a common concept, which is not explicit in the code - a kind of array provider (currently, stateless with arguments, or stateful without arguments) that is internally modeled by an "init" and an "execute" functions. I think both views are valid, and I take the latter view which is the easier one.

> It smells a little bit to me of "using a hammer because you have one lying around and not because its the best tool for the job"

I believe this smell is an artifact of the first view and goes away with the second view. In particular, I think that if we decided to define a new kind of function category with the second view in mind, we'd probably be content with an "init" and "execute" functions anyway.

>  I suspect you may run into trouble once you want to start passing in arguments (you'd want those to go to the init function right? I'm not sure how you can do that).

A related discussion is [here](https://github.com/apache/arrow/pull/14682/#discussion_r1029168570).

UDT arguments wouldn't be passed to the "init" nor to the "execute" function but to the registered UDT function (like `udf_func` in the tester) returned by the Python UDT maker (like `make_udt_func` in the tester). These arguments, or values derived from them, can be saved on the returned UDT instance.

Reply:
Regarding `StopIteration`, since it is handled by the `UDT` class, perhaps what we'd want is to make (something like) this class part of the API so that the user would only need to provide the generator.

Reply:
> Regarding StopIteration, since it is handled by the UDT class, perhaps what we'd want is to make (something like) this class part of the API so that the user would only need to provide the generator.

Agreed.  That would be a good solution.  It can wait for a future PR if needed.

> UDT arguments wouldn't be passed to the "init" nor to the "execute" function but to the registered UDT function (like udf_func in the tester) returned by the Python UDT maker (like make_udt_func in the tester). These arguments, or values derived from them, can be saved on the returned UDT instance.

First, to clarify, by "UDT arguments" I am referring to the input rows (i.e. a UDT with non-null arity).  I think that is what you are referring to as well.  I agree that these args will need to be passed to the equivalent of `udf_func`.  However, `udf_func` is currently called by the kernel's init function.

In other words, if I add these prints:

```
// udf.cc CallTabularFunction
  std::cout << "About to get executor" << std::endl;
  ARROW_ASSIGN_OR_RAISE(auto func_exec,
                        GetFunctionExecutor(func_name, in_types, NULLPTR, registry));
  std::cout << "Retrieved executor" << std::endl;

// test_udf.py
    def udf_func(ctx):
        class UDT:
            ...
        print("Creating UDT")
        return UDT()
```

Then I see the output:

```
About to get executor
Creating UDT
Retrieved executor
```

The problem is that a kernel's init function does not accept arguments today (e.g. you can't have a unary or binary init function).  So you will need to solve this problem when you're ready to create UDTs that have non-null arity.

Reply:
> First, to clarify, by "UDT arguments" I am referring to the input rows (i.e. a UDT with non-null arity). I think that is what you are referring to as well.

I think we're on the same page here, though I actually had in mind simpler arguments, like numbers or strings, rather than (a table with) input rows, which I guess is also possible.

> The problem is that a kernel's init function does not accept arguments today (e.g. you can't have a unary or binary init function). So you will need to solve this problem when you're ready to create UDTs that have non-null arity.

Though I haven't validated this, the way I think it could work is via `KernelInitArgs`. That is, the arguments would be placed in an appropriate `FunctionOptions` and passed to `FunctionExecutor::Init`, which would deliver them to `Kernel::Init`. This would arrive at `PythonTableUdfKernelInit::operator()(compute::KernelContext* ctx, const compute::KernelInitArgs& init_args)`, whose logic would be changed to extract arguments from `init_args` and place them in a tuple, replacing the current `empty_tuple` in the PR code. These arguments would then be passed via the `cb` callback as extra argument (in addition to `ctx`) to `udf_func`, which would bake them into the UDT instance it returns.

Reply:
> Agreed. That would be a good solution. It can wait for a future PR if needed.

https://github.com/apache/arrow/issues/33879